sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java

```diff
@@ -0,0 +1,735 @@
+     */
+    public Read<T> fromQuery(String query) {
+      return toBuilder().setQuery(query).build();
+    }
+
+    /**
+     * A table name to be read in Snowflake.
+     *
+     * @param table - String with the name of the table.
+     */
+    public Read<T> fromTable(String table) {
+      return toBuilder().setTable(table).build();
+    }
+
+    /**
+     * Name of the cloud bucket (GCS by now) to use as tmp location of CSVs during COPY statement.
+     *
+     * @param stagingBucketName - String with the name of the bucket.
+     */
+    public Read<T> withStagingBucketName(String stagingBucketName) {
+      return toBuilder().setStagingBucketName(stagingBucketName).build();
+    }
+
+    /**
+     * Name of the Storage Integration in Snowflake to be used. See
+     * https://docs.snowflake.com/en/sql-reference/sql/create-storage-integration.html for
+     * reference.
+     *
+     * @param integrationName - String with the name of the Storage Integration.
+     */
+    public Read<T> withIntegrationName(String integrationName) {
+      return toBuilder().setIntegrationName(integrationName).build();
+    }
+
+    /**
+     * User-defined function mapping CSV lines into user data.
+     *
+     * @param csvMapper - an instance of {@link CsvMapper}.
+     */
+    public Read<T> withCsvMapper(CsvMapper<T> csvMapper) {
+      return toBuilder().setCsvMapper(csvMapper).build();
+    }
+
+    /**
+     * A Coder to be used by the output PCollection generated by the source.
+     *
+     * @param coder - an instance of {@link Coder}.
+     */
+    public Read<T> withCoder(Coder<T> coder) {
+      return toBuilder().setCoder(coder).build();
+    }
+
+    @Override
+    public PCollection<T> expand(PBegin input) {
+      // Either table or query is required. If query is present, it's being used, table is used
+      // otherwise
+      checkArgument(
+          getQuery() != null || getTable() != null, "fromTable() or fromQuery() is required");
+      checkArgument(
+          !(getQuery() != null && getTable() != null),
+          "fromTable() and fromQuery() are not allowed together");
+      checkArgument(getCsvMapper() != null, "withCsvMapper() is required");
+      checkArgument(getCoder() != null, "withCoder() is required");
+      checkArgument(getIntegrationName() != null, "withIntegrationName() is required");
+      checkArgument(getStagingBucketName() != null, "withStagingBucketName() is required");
+      checkArgument(
+          (getDataSourceProviderFn() != null),
+          "withDataSourceConfiguration() or withDataSourceProviderFn() is required");
+
+      String gcpTmpDirName = makeTmpDirName();
+      String stagingBucketDir = String.format("%s/%s/", getStagingBucketName(), gcpTmpDirName);
+
+      PCollection<Void> emptyCollection = input.apply(Create.of((Void) null));
+
+      PCollection<T> output =
+          emptyCollection
+              .apply(
+                  ParDo.of(
+                      new CopyIntoStageFn(
```

Comment:
What will happen in case of a failure and a retry of a bundle ?

I think you should add a Reshuffle step after CopyIntoStageFn() step to checkpoint the set of files (supported by most runners).

Also, make sure you have considered what would happen if CopyIntoStageFn() itself fails and retries. For example, BQ source tries to look for an already started export job before starting a new one.

Reply:
If running `COPY` fails and it will be retried it will fail on any retry. 
It's because the first attempt will put some files into staging bucket and another try will find out that staging location is not empty and it will fail.

There is a `OVERWRITE` option in Snowflake, telling that if there are files on cloud storage they will be overwritten, but we don't use it. The reason is simple - we're reading all data from this temporary location. If for any reason, on this location there will be garbage files, data will be malformed. I imagine it might happen that first attempt will put some files, and another attempt might have slightly different filenames and files will be appended. In this case we have partial results from first try and full results from second - we don't want that.

There is also no such thing as "job" in this case. It's just, simple, standalone "plain old good sql query" that is executed against snowflake. It is parallelised, but internally, on Snowflake side - from pipeline perspective it's one-off task. It doesn't have a job_id which status can be checked and resumed or waited for finish.

I'd suggest to keep it as is - failed pipeline in this case might be better solution than broken data. WDYT?

Reply:
It's better but I don't think it's a good user experience to fail the pipeline whenever a bundle is retried. Bundle retries are a usual part of the runner execution and source/sink transforms should be able to handle that. 

What is usually done in this case is to write data to a new temporary location each time the copy step is run. Then you add a Reshuffle write after the copy operation to checkpoint the results. This way you can guarantee that the set of files that are output to the next step are from a single execution and the job does not have to fail simply because a bundle is retried.

When you cleanup you have to clean all temporary locations including retries.

I'm fine if you don't want to handle this in the first try but let's at least add a TODO and a JIRA to fix this later.

Reply:
Thanks @chamikaramj for insisting on this. The approach you suggested works perfectly and was easy to incorporate into our case.

I've made changes that are creating a new sub-directory for each attempt of `CopyIntoStageFn`. If this Fn is retried, another subdirectory is created and if step succeeds, exactly this subdirectory will be passed to `FileIO` to further processing. In the end, all sub-directories are cleaned-up.

----------------

I'm however not sure about `Reshuffle`. In our case output from `CopyIntoStageFn` is single String containing cloud storage path from where files should be read by FileIO. Does it make sense to reshuffle it? I've tested it and it works well with retries without reshuffle.
Also, Reshuffle is deprecated and might not work well in portability (if I understand docs properly) and we're aiming to use cross-language to support Python SDK.
Do you think the `Reshuffle` step is needed here?

Reply:
So you have following two steps that will get fused by the runner and run together.
ParDo(CopyIntoStageFn)+ParDo(ReadFiles)

Now assume there's a failure when reading files. This will result in expensive CopyIntoStageFn from being executed again.

Reshuffle (for most runners) has the affect of breaking fusion and hence checkpointing results of ParDo(CopyIntoStageFn). So if there's a failure when reading, the ParDo(CopyIntoStageFn) step will not have to be re-run. Also this is a very inexpensive shuffle (we only send file-names through it). 

Reshuffle is deprecated since its behavior across runners is not well defined. Some runners may consider it to be an identity operation while most runners will checkpoint results. I don't think it can be removed from the SDK anytime soon and many widely used connectors depend on it.

Reply:
Thanks @chamikaramj for detailed explanation. I've added `Reshuffle` step in [latest commit](https://github.com/apache/beam/pull/11360/commits/3ba192a57217a9b251f71f77771b68ceb2882b02)